### PR TITLE
bpo-43923: Allow NamedTuple multiple inheritance

### PR DESCRIPTION
**Pull request by juliusgeo**
https://bugs.python.org/issue43923

This is a fix-forward PR (partial revert of [bpo-40185](https://bugs.python.org/issue40185)).

https://bugs.python.org/issue43923

### CODE REVIEW

File: Lib/typing.py [hunk 1]
```diff
@@ -2595,13 +2595,15 @@ def __round__(self, ndigits: int = 0) -> T_co:
...
-    fields = [n for n, t in types]
-    types = {n: _type_check(t, f"field {n} annotation must be a type")
-             for n, t in types}
-    nm_tpl = collections.namedtuple(name, fields,
-                                    defaults=defaults, module=module)
-    nm_tpl.__annotations__ = nm_tpl.__new__.__annotations__ = types
+def _make_nmtuple(name, types):
+    msg = "NamedTuple('Name', [(f0, t0), (f1, t1), ...]); each t must be a type"
+    types = [(n, _type_check(t, msg)) for n, t in types]
+    nm_tpl = collections.namedtuple(name, [n for n, t in types])
```
Comment (JelleZijlstra) [inline Lib/typing.py]: `collections.namedtuple` now supports `default=` directly, so we should use that

File: Lib/typing.py [hunk 2]
```diff
@@ -2595,13 +2595,15 @@ def __round__(self, ndigits: int = 0) -> T_co:
...
 
-def _make_nmtuple(name, types, module, defaults = ()):
-    fields = [n for n, t in types]
-    types = {n: _type_check(t, f"field {n} annotation must be a type")
-             for n, t in types}
-    nm_tpl = collections.namedtuple(name, fields,
-                                    defaults=defaults, module=module)
-    nm_tpl.__annotations__ = nm_tpl.__new__.__annotations__ = types
+def _make_nmtuple(name, types):
+    msg = "NamedTuple('Name', [(f0, t0), (f1, t1), ...]); each t must be a type"
```
Comment (JelleZijlstra) [inline Lib/typing.py]: You're missing the improved error message from the existing code

File: Lib/typing.py [hunk 3]
```diff
@@ -2616,20 +2618,26 @@ def _make_nmtuple(name, types, module, defaults = ()):
...
+                defaults.append(default_value)
+                defaults_dict[field_name] = default_value
+            elif defaults:
+                raise TypeError("Non-default namedtuple field {field_name} cannot "
+                                "follow default field(s) {default_names}"
+                                .format(field_name=field_name,
+                                        default_names=', '.join(defaults_dict.keys())))
+        nm_tpl.__new__.__annotations__ = dict(types)
+        nm_tpl.__new__.__defaults__ = tuple(defaults)
+        nm_tpl._field_defaults = defaults_dict
```
Comment (JelleZijlstra) [inline Lib/typing.py]: No need to restore this old attribute

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
This looks like it's still mostly a revert, which I don't think is the right approach since it ends up losing a bunch of past improvements. Try to make the minimal change necessary to make things work.

**Comment by JelleZijlstra** [comment]
Note there's an alternative PR from Serhyi: #31781.

**Comment by AlexWaygood** [comment]
> Note there's an alternative PR from Serhiy: #31781.

I still like Serhiy's more. In particular, I think it's best to keep `NamedTuple` a function, rather than a class. Given what inheriting from `NamedTuple` actually does (dynamically create a new _tuple_ subclass), I think it makes much more sense for it to be a function.

**Comment by JelleZijlstra** [comment]
We merged #92027 to allow generic namedtuples. I don't think we want more general multiple inheritance. Thanks for your PR though!

**Pull request closed without merging**